arch/sim/src/sim/up_initialstate.c

```diff
@@ -86,7 +86,13 @@ void up_initial_state(struct tcb_s *tcb)
 
   tcb->xcp.regs[JB_SP] = (xcpt_reg_t)tcb->stack_base_ptr +
                                      tcb->adj_stack_size -
+#if defined(CONFIG_HOST_ARM64)
+  /* SP needs 16byte alignment on ARM64 architecture */
```

Comment:
let's correct the alignment in first place instead(up_create_stack/up_use_stack/up_stack_frame).

Reply:
That's not possible. The issues is exactly this place. tcb->stack_base_ptr and tcb->adj_stack_size are correctly 16 byte aligned. But this code allocates a single 64bit, resulting in tcb->xcp.regs[JB_SP] gets 8byte aligned.

Reply:
Do you know why we need allocate 8/16 bytes here?

Reply:
It's part of the ARM64 architecture itself and explicitly required on Apple platforms: https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms --> "Pass Arguments to Functions Correctly"

Reply:
Look like the link doesn't answer my question: why we need sub sizeof(xcpt_reg_t) or 2 * sizeof(xcpt_reg_t) from sp.

Reply:
From the PDF:
  Pass Arguments to Functions Correctly
  The **stack pointer on Apple platforms follows the ARM64 standard ABI and requires 16-byte alignment**.
That's why 2*sizeof(xcpt_reg_t) as sizeof(xcpt_reg_t) is 8 bytes only.

Reply:
Yes, I know the alignment requirement, but I don't understand why we need subtract.

Reply:
Got it ... honestly spoken ... I don't. The subtract was part of the code already. Directly in front of the code, there's a comment, explaining the reason. I took this for granted and adapted it to the ARM64 requirements as the simulation was throwing an exception without it.

     /* Note: The amd64 ABI requires 16-bytes alignment _before_ a function
      * call.
      * On the other hand, our way to set up and switch to a new context
      * is basically a JUMP.
      * Thus, we need to emulate the effect of a CALL here, by subtracting
      * sizeof(xcpt_reg_t), which is the amount a CALL would move RSP to store
      * the return address.
      */

Reply:
From the description, I think setjmp/longjmp is the better place to fix this issue. Could you try to fix this issue there instead?

Reply:
Overall subtract or just the alignment for ARM64?

Reply:
Both, setjmp/longjmp should handle the alignment and subtract correctly.

Reply:
I get your point, but in my opinion it's not the best solution to calculate a wrongly aligned stack pointer in up_initial_state() and let setjmp/longjmp later correct it. To avoid the ifdef and ugly subtract, I changed the code to calculate the initial stack pointer in a different way. New calculation leaves one "alignment size" (16 byte) free at the end of the stack to simulate what the comment describes.

Reply:
The change look like much better than before, but I still have the same question: if up_create_stack/up_use_stack/up_stack_frame already ensure the alignment correctly, why we need align again in up_initial_state?

Reply:
Ok, after review the related code again, the root cause is that up_create_stack/up_use_stack doesn't allocate/adjust the stack to start at 16 bytes boundary. This is common issue for both x64 and arm64, could you create a separated patch fix this issue? You can reference the similar arm code here:
https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/common/arm_createstack.c
https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/common/arm_usestack.c
https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/common/arm_stackframe.c
The subtraction in up_initial_state could be removed, just like other arch:
https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/armv8-m/arm_initialstate.c

Reply:
ping @ptka

Reply:
I reviewed code again and the subtraction cannot be removed. There is a specific reason for it (not related to any aligment issue) on the amd64 architecture:

    /* Note: The amd64 ABI requires 16-bytes alignment _before_ a function
     * call.
     * On the other hand, our way to set up and switch to a new context
     * is basically a JUMP.
     * Thus, we need to emulate the effect of a CALL here, by subtracting
     * sizeof(xcpt_reg_t), which is the amount a CALL would move RSP to store
     * the return address.
     */

The subtraction simulates how a CALL would change the stack.

On the ARM64 architecture the return address is not stored on the stack, but in LR register. There's no need for the subtraction on the ARM64 architecture as the stack is unchanged.

I changed the code in a way, that the subtraction is only effective on none ARM64 architectures and added a note.

     * Note: On ARM64 architectures the return address is passed via LR
     *       register. No extra adjustment for the stack needed.